### ISSUE TITLE
Siege socket errors

### ISSUE THREAD

**Issue by keshonok** [natsys/tempesta#376]
Now that client connections can be closed by `Tempesta` in certain cases (see #348, #361), I see lots of sockets in `TIME_WAIT` state. That happens on the side with a back end server (such as `Apache`), and on the side with a client (such as `siege` being a benchmark tool). At certain moment all sockets get exhausted, the server stops responding to connection attempts, and the client finds that it can't open a new socket.

At some point I made an attempt to increase the number of open files in the `siege` client session but that didn't help much.

On `siege` client side:
```
lta@bastinda:~/tfw-run/wget$ ulimit -Hn
65535
lta@bastinda:~/tfw-run/wget$ ulimit -Sn
65535
```

On `siege` client side:
```
[error] socket: -1331792128 address is unavailable.: Cannot assign requested address
[error] socket: -1306614016 address is unavailable.: Cannot assign requested address
[error] socket: -1289828608 address is unavailable.: Cannot assign requested address
[error] socket: -1298221312 address is unavailable.: Cannot assign requested address
[error] socket: -1231079680 address is unavailable.: Cannot assign requested address
done.
siege aborted due to excessive socket failure; you
can change the failure threshold in $HOME/.siegerc
```

On `siege` client side:
```
lta@bastinda:~/tfw-run/wget$ ss -s
Total: 766 (kernel 0)
TCP:   28270 (estab 18, closed 28194, orphaned 0, synrecv 0, timewait 28190/0), ports 0

Transport Total     IP        IPv6
*         0         -         -        
RAW       0         0         0        
UDP       15        13        2        
TCP       76        70        6        
INET      91        83        8        
FRAG      0         0         0        
```

on `Apache` back end side:
```
[root@localhost ~]# ss -s
Total: 425 (kernel 450)
TCP:   16417 (estab 25, closed 16386, orphaned 4, synrecv 0, timewait 16384/0), ports 41

Transport Total     IP        IPv6
*         450       -         -        
RAW       1         0         1        
UDP       9         6         3        
TCP       31        24        7        
INET      41        30        11       
FRAG      0         0         0        
```

**Comment by keshonok** [comment]
This issue has regressed significantly after 1d29b54. Not the `siege` client stops nearly immediately due to excessive socket failure, as obviously it gets lots of error responses from `Tempesta`.

```
HTTP/1.1 404   0.01 secs:       0 bytes ==> GET  /
HTTP/1.1 404   0.00 secs:       0 bytes ==> GET  /
HTTP/1.1 404   0.01 secs:       0 bytes ==> GET  /
HTTP/1.1 404   0.00 secs:       0 bytes ==> GET  /
HTTP/1.1 404   0.01 secs:       0 bytes ==> GET  /
HTTP/1.1 404   0.00 secs:       0 bytes ==> GET  /
HTTP/1.1 404   0.00 secs:       0 bytes ==> GET  /
HTTP/1.1 404   0.00 secs:       0 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.01 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.01 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.01 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
HTTP/1.1 200   0.00 secs:      46 bytes ==> GET  /
done.
siege aborted due to excessive socket failure; you
can change the failure threshold in $HOME/.siegerc

Transactions:                    293 hits
Availability:                  22.01 %
Elapsed time:                   0.14 secs
Data transferred:               0.00 MB
Response time:                  0.00 secs
Transaction rate:            2092.86 trans/sec
Throughput:                     0.02 MB/sec
Concurrency:                    2.71
Successful transactions:          63
Failed transactions:            1038
Longest transaction:            0.04
Shortest transaction:           0.00
```

**Comment by keshonok** [comment]
The original topic name appears to be not relevant to the problem that occurs. Even with the number of sockets in TIME_WAIT state up to the maximum, benchmarks and web servers work continuously and without stopping.

This particular problem occurs when running `siege` in benchmark mode against Tempesta. It doesn't matter if the `connection` option in `.siegerc` is set to default `close` or if it is set to `keep-alive`. `siege` runs on the host machine, while Tempesta, Nginx, and Apache all run in a VM.

Exactly the same symptoms are exhibited when `siege` is run against Apache in default configuration. Running `siege` again Nginx **does not** exhibit this behaviour.

After a short run of about 30 seconds a `siege` test stops with the following symptoms:
```
[error] socket: -325073152 address is unavailable.: Cannot assign requested address
[error] socket: -257931520 address is unavailable.: Cannot assign requested address
[error] socket: -249538816 address is unavailable.: Cannot assign requested address
[error] socket: -224360704 address is unavailable.: Cannot assign requested address
[error] socket: -241146112 address is unavailable.: Cannot assign requested address
[error] socket: -299895040 address is unavailable.: Cannot assign requested address
[error] socket: -308287744 address is unavailable.: Cannot assign requested address
[error] socket: -274716928 address is unavailable.: Cannot assign requested address
[error] socket: -266324224 address is unavailable.: Cannot assign requested address
[error] socket: -215968000 address is unavailable.: Cannot assign requested address
[error] socket: -207575296 address is unavailable.: Cannot assign requested address
[error] socket: -283109632 address is unavailable.: Cannot assign requested address
[error] socket: -291502336 address is unavailable.: Cannot assign requested address
done.
siege aborted due to excessive socket failure; you
can change the failure threshold in $HOME/.siegerc
```

Apache's `ab` benchmark doesn't induce this kind of behaviour from Tempesta, Apache, or Nginx.

Tracing shows that Apache and Nginx process incoming requests in a different manner. Nginx uses TCP corking, while Apache does not.

Nginx:
```
[pid  1153] recvfrom(8, "GET / HTTP/1.1\r\nUser-Agent: Wget"..., 1024, 0, NULL, NULL) = 117
[pid  1153] stat("/usr/share/nginx/html/index.html", {st_dev=makedev(253, 0), st_ino=3684008, st_mode=S_IFREG|0644, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=3700, st_atime=2016/02/03-14:05:48, st_mtime=2015/10/04-10:53:44, st_ctime=2015/12/28-03:44:30}) = 0
[pid  1153] open("/usr/share/nginx/html/index.html", O_RDONLY|O_NONBLOCK) = 9
[pid  1153] fstat(9, {st_dev=makedev(253, 0), st_ino=3684008, st_mode=S_IFREG|0644, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=3700, st_atime=2016/02/03-14:05:48, st_mtime=2015/10/04-10:53:44, st_ctime=2015/12/28-03:44:30}) = 0
[pid  1153] setsockopt(8, SOL_TCP, TCP_CORK, [1], 4) = 0
[pid  1153] writev(8, [{"HTTP/1.1 200 OK\r\nServer: nginx/1"..., 238}], 1) = 238
[pid  1153] sendfile(8, 9, [0], 3700)   = 3700
[pid  1153] write(5, "192.168.10.1 - - [03/Feb/2016:14"..., 104) = 104
[pid  1153] close(9)                    = 0
[pid  1153] setsockopt(8, SOL_TCP, TCP_CORK, [0], 4) = 0
[pid  1153] recvfrom(8, 0x55f7e2a21f60, 1024, 0, 0, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid  1153] epoll_wait(11, {{EPOLLIN|EPOLLRDHUP, {u32=1600087545, u64=139643871782393}}}, 512, 65000) = 1
[pid  1153] recvfrom(8, "", 1024, 0, NULL, NULL) = 0
[pid  1153] close(8)                    = 0
```

Apache:
```
[pid  2298] select(0, NULL, NULL, NULL, {1, 0} <unfinished ...>
[pid  2317] <... accept4 resumed> {sa_family=AF_INET6, sin6_port=htons(57453), inet_pton(AF_INET6, "::ffff:192.168.10.1", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28], SOCK_CLOEXEC) = 9
[pid  2317] getsockname(9, {sa_family=AF_INET6, sin6_port=htons(8080), inet_pton(AF_INET6, "::ffff:192.168.10.230", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28]) = 0
[pid  2317] fcntl(9, F_GETFL)           = 0x2 (flags O_RDWR)
[pid  2317] fcntl(9, F_SETFL, O_RDWR|O_NONBLOCK) = 0
[pid  2317] read(9, "GET / HTTP/1.1\r\nUser-Agent: Wget"..., 8000) = 117
[pid  2317] stat("/var/www/html/", {st_dev=makedev(253, 0), st_ino=113407026, st_mode=S_IFDIR|0755, st_nlink=2, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=0, st_size=23, st_atime=2015/11/20-00:43:57, st_mtime=2015/11/20-00:43:57, st_ctime=2016/02/03-13:48:33}) = 0
[pid  2317] stat("/var/www/html/index.html", {st_dev=makedev(253, 0), st_ino=113407035, st_mode=S_IFREG|0644, st_nlink=1, st_uid=0, st_gid=0, st_blksize=4096, st_blocks=8, st_size=46, st_atime=2016/02/03-13:17:02, st_mtime=2015/11/20-11:45:05, st_ctime=2015/11/20-11:45:05}) = 0
[pid  2317] open("/var/www/html/index.html", O_RDONLY|O_CLOEXEC) = 10
[pid  2317] read(9, 0x55ae241d1088, 8000) = -1 EAGAIN (Resource temporarily unavailable)
[pid  2317] mmap(NULL, 46, PROT_READ, MAP_SHARED, 10, 0) = 0x7ffb524eb000
[pid  2317] writev(9, [{"HTTP/1.1 200 OK\r\nDate: Wed, 03 F"..., 289}, {"This is a sample index file.\nJus"..., 46}], 2) = 335
[pid  2317] munmap(0x7ffb524eb000, 46)  = 0
[pid  2317] write(7, "192.168.10.1 - - [03/Feb/2016:14"..., 98) = 98
[pid  2317] times({tms_utime=7, tms_stime=18, tms_cutime=0, tms_cstime=0}) = 429539563
[pid  2317] close(10)                   = 0
[pid  2317] poll([{fd=9, events=POLLIN}], 1, 60000) = 1 ([{fd=9, revents=POLLIN}])
[pid  2317] read(9, "", 8000)           = 0
[pid  2317] shutdown(9, SHUT_WR)        = 0
[pid  2317] poll([{fd=9, events=POLLIN}], 1, 2000) = 1 ([{fd=9, revents=POLLIN|POLLHUP}])
[pid  2317] read(9, "", 512)            = 0
[pid  2317] close(9)                    = 0
[pid  2317] read(5, 0x7ffcc1ff024f, 1)  = -1 EAGAIN (Resource temporarily unavailable)
[pid  2317] accept4(4,  <unfinished ...>
[pid  2298] <... select resumed> )      = 0 (Timeout)
```

**Comment by krizhanovsky** [comment]
Actually ever just after Tempesta start there is socket issue:

        # netstat -npta|grep 8080
        tcp        0      0 127.0.0.1:8080          127.0.0.1:60528         SYN_RECV    -                   
        tcp        0      0 127.0.0.1:8080          127.0.0.1:60530         SYN_RECV    -                   
        tcp        0      0 127.0.0.1:8080          127.0.0.1:60531         SYN_RECV    -                   
        tcp        0      0 127.0.0.1:8080          127.0.0.1:60529         SYN_RECV    -                   
        tcp        0      0 127.0.0.1:60529         127.0.0.1:8080          ESTABLISHED -                   
        tcp        0      0 127.0.0.1:60531         127.0.0.1:8080          ESTABLISHED -                   
        tcp        0      0 127.0.0.1:60530         127.0.0.1:8080          ESTABLISHED -                   
        tcp        0      0 127.0.0.1:60528         127.0.0.1:8080          ESTABLISHED -                   
        tcp6       0      0 :::8080                 :::*                    LISTEN      794/httpd  

It seems we don't finish TCP handshake. Probably not related to original issue, but it seems this is very close thing.

**Comment by keshonok** [comment]
This happens because Apache uses `TCP_DEFER_ACCEPT` socket option. So right after Tempesta is started with `Apache` as a back end server, the sockets in `Apache` are in `SYN_RECV` state. As soon as data is detected in these sockets, the sockets are moved to `TCP_ESTABLISHED` state.

Nginx doesn't exhibit this behaviour.

**Comment by keshonok** [comment]
Attached are the configuration file for `Apache httpd`, and the configuration file for `siege`. The changes to the default configurations of both programs are minimal.

[httpd.conf.txt](https://github.com/natsys/tempesta/files/149985/httpd.conf.txt)
[.siegerc.txt](https://github.com/natsys/tempesta/files/149987/default.siegerc.txt)

**Comment by krizhanovsky** [comment]
The problem is hard to reproduce on my setup. However, I saw it twice with following symptoms:

1. my own, bit different, Apache HTTPD and siege configs (siege was run with concurrency 32);

2. reproduced against Tempesta a front of Apache HTTPD as reverse proxy and Apache HTTPD stand alone;

3. tcpdump shows that siege doesn't send FIN in response to Tempesta's or Apache HTTPD's FIN, i.e. there are a lot of connections in `FIN_WAIT_2` state. This is the first problem with siege. However, Alexey's dumps don't show such siege behavior;

4. siege ends with messages (see also this [old bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=595527)):

        [error] descriptor table full sock.c:132: Too many open files
        [error] descriptor table full sock.c:132: Too many open files
        libgcc_s.so.1 must be installed for pthread_cancel to work

This is the second problem: with specified concurrency to 32 parallel connections, default ulimit as of 1024 file descriptors is not enough. So Alexey increased the limits to 64K. And with the limits siege throws `address is unavailable` error. This is `EADDRNOTAVAIL` return value from `connect(2)` system call, `man 2 connect` says:

        EADDRNOTAVAIL
              (Internet  domain  sockets)  The  socket  referred to by sockfd had not previously
              been bound to an address and, upon attempting to bind it to an ephemeral port,
              it was determined that all port  numbers in the ephemeral port range are
              currently in use.

So the 3rd problem is that siege eats all available ports on the client system.

Meantime, the issue doesn't exhibit any Tempesta misbehavior and the original problem is fully in not enough system resources at client side. So we just don't recommend to use siege. `ab` and `wrk` didn't show such issues.

However, during the tests we found that the test goes smoothly with Nginx. The reason is that Nginx immediately places FIN at last data chunk, together with PUSH flag, so connection termination goes quicker and siege doesn't eat all the system ports and file descriptors. So this is a good behavior and we place it in our TODO.
